rust/src/lib/nm/show.rs

```diff
@@ -155,52 +168,57 @@ pub(crate) fn nm_retrieve() -> Result<NetworkState, NmstateError> {
     Ok(net_state)
 }
 
-fn nm_iface_type_to_nmstate(nm_iface_type: &str) -> InterfaceType {
-    match nm_iface_type {
+fn nm_dev_iface_type_to_nmstate(nm_dev: &NmDevice) -> InterfaceType {
+    match nm_dev.iface_type.as_str() {
         NM_SETTING_WIRED_SETTING_NAME => InterfaceType::Ethernet,
         NM_SETTING_VETH_SETTING_NAME => InterfaceType::Ethernet,
         NM_SETTING_BOND_SETTING_NAME => InterfaceType::Bond,
         NM_SETTING_DUMMY_SETTING_NAME => InterfaceType::Dummy,
         NM_SETTING_BRIDGE_SETTING_NAME => InterfaceType::LinuxBridge,
         NM_SETTING_OVS_BRIDGE_SETTING_NAME => InterfaceType::OvsBridge,
         NM_SETTING_OVS_IFACE_SETTING_NAME => InterfaceType::OvsInterface,
-        _ => InterfaceType::Other(nm_iface_type.to_string()),
+        NM_SETTING_MACVLAN_SETTING_NAME => {
+            if nm_dev.is_mac_vtap {
+                InterfaceType::MacVtap
+            } else {
+                InterfaceType::MacVlan
+            }
+        }
+        _ => InterfaceType::Other(nm_dev.iface_type.to_string()),
     }
 }
 
-fn nm_conn_to_base_iface(nm_conn: &NmConnection) -> Option<BaseInterface> {
+fn nm_conn_to_base_iface(
+    nm_dev: &NmDevice,
+    nm_conn: &NmConnection,
+) -> Option<BaseInterface> {
     if let Some(iface_name) = nm_conn.iface_name() {
-        if let Some(iface_type) = nm_conn.iface_type() {
-            let ipv4 = nm_conn.ipv4.as_ref().map(|nm_ipv4_setting| {
-                nm_ip_setting_to_nmstate4(nm_ipv4_setting)
-            });
-            let ipv6 = nm_conn.ipv6.as_ref().map(|nm_ipv6_setting| {
-                nm_ip_setting_to_nmstate6(nm_ipv6_setting)
-            });
+        let ipv4 = nm_conn.ipv4.as_ref().map(nm_ip_setting_to_nmstate4);
+        let ipv6 = nm_conn.ipv6.as_ref().map(nm_ip_setting_to_nmstate6);
 
-            let mut base_iface = BaseInterface::new();
-            base_iface.name = iface_name.to_string();
-            base_iface.prop_list =
-                vec!["name", "state", "iface_type", "ipv4", "ipv6"];
-            base_iface.state = InterfaceState::Up;
-            base_iface.iface_type = nm_iface_type_to_nmstate(iface_type);
-            base_iface.ipv4 = ipv4;
-            base_iface.ipv6 = ipv6;
-            base_iface.controller = nm_conn.controller().map(|c| c.to_string());
-            return Some(base_iface);
-        }
+        let mut base_iface = BaseInterface::new();
+        base_iface.name = iface_name.to_string();
+        base_iface.prop_list =
+            vec!["name", "state", "iface_type", "ipv4", "ipv6"];
+        base_iface.state = InterfaceState::Up;
+        base_iface.iface_type = nm_dev_iface_type_to_nmstate(nm_dev);
+        base_iface.ipv4 = ipv4;
+        base_iface.ipv6 = ipv6;
+        base_iface.controller = nm_conn.controller().map(|c| c.to_string());
+        return Some(base_iface);
```

Comment:
I agree with these changes but I do not think they are related to this commit. Could you move this to its own commit. You can keep it on this PR IMO.

Reply:
actually, they are related to this PR. This is the new way on checking interface type, previously we just use `nm_iface_type`, but for MacVtap, we need `nm_dev` to detemine whether is it MacVlan or MacVtap.

Anyway, I will split it out.

Reply:
Patch isolated.